### ISSUE TITLE
[Failing test] GCE-5000: additional list calls in load test

### ISSUE THREAD

**Issue by krzysied** [kubernetes/kubernetes#79096]
**Which jobs are failing**:
gce-scale-performance

**Which test(s) are failing**:
Laod test

**Since when has it been failing**:
2019-05-28 10:01 CEST (?)

**Testgrid link**:
https://testgrid.k8s.io/sig-scalability-gce#gce-scale-performance

**Reason for failure**:
There are additional services, pods, nodes calls made by kubelets. Test not always fails due to this, however those call have impact on other calls latency.

**Anything else we need to know**:

**Comment by krzysied** [comment]
![kubelet_services_list](https://user-images.githubusercontent.com/25119199/59593525-67d4de80-90f2-11e9-8109-9b600d3be722.png)
[services list - kubelet](http://perf-dash.k8s.io/#/?jobname=gce-5000Nodes&metriccategoryname=APIServer&metricname=LoadRequestCountByClient&client=kubelet%2Fv1.16.0%20(linux%2Famd64)%20kubernetes&code=200&component=apiserver&group=&resource=services&scope=cluster&subresource=&verb=LIST&version=v1)

[nodes list - kubelet](http://perf-dash.k8s.io/#/?jobname=gce-5000Nodes&metriccategoryname=APIServer&metricname=LoadRequestCountByClient&client=kubelet%2Fv1.16.0%20(linux%2Famd64)%20kubernetes&code=200&component=apiserver&group=&resource=nodes&scope=cluster&subresource=&verb=LIST&version=v1)

[pods list - kubelet](http://perf-dash.k8s.io/#/?jobname=gce-5000Nodes&metriccategoryname=APIServer&metricname=LoadRequestCountByClient&client=kubelet%2Fv1.16.0%20(linux%2Famd64)%20kubernetes&code=200&component=apiserver&group=&resource=pods&scope=cluster&subresource=&verb=LIST&version=v1)

**Comment by krzysied** [comment]
```
W0613 17:07:29.494211    1489 reflector.go:302] object-"test-y8pkr4-16"/"default-token-qqt29": watch of *v1.Secret ended with: very short watch: object-"test-y8pkr4-16"/"default-token-qqt29": Unexpected watch close - watch lasted less than a second and no items received
W0613 17:07:29.494752    1489 reflector.go:302] object-"test-y8pkr4-38"/"default-token-q58n8": watch of *v1.Secret ended with: very short watch: object-"test-y8pkr4-38"/"default-token-q58n8": Unexpected watch close - watch lasted less than a second and no items received
W0613 17:07:29.496888    1489 reflector.go:302] object-"kube-system"/"fluentd-gcp-token-kzzzv": watch of *v1.Secret ended with: very short watch: object-"kube-system"/"fluentd-gcp-token-kzzzv": Unexpected watch close - watch lasted less than a second and no items received
W0613 17:07:29.497198    1489 reflector.go:302] object-"test-y8pkr4-9"/"default-token-p2wvb": watch of *v1.Secret ended with: very short watch: object-"test-y8pkr4-9"/"default-token-p2wvb": Unexpected watch close - watch lasted less than a second and no items received
W0613 17:07:29.497487    1489 reflector.go:302] k8s.io/kubernetes/pkg/kubelet/kubelet.go:453: watch of *v1.Node ended with: very short watch: k8s.io/kubernetes/pkg/kubelet/kubelet.go:453: Unexpected watch close - watch lasted less than a second and no items received
W0613 17:07:29.498884    1489 reflector.go:302] k8s.io/kubernetes/pkg/kubelet/config/apiserver.go:47: watch of *v1.Pod ended with: very short watch: k8s.io/kubernetes/pkg/kubelet/config/apiserver.go:47: Unexpected watch close - watch lasted less than a second and no items received
```
Based on kubelet logs, it looks like apiserver is closing watch connection. To recreate watches kubelet needs to list resources. This would explain these additional calls.

**Comment by krzysied** [comment]
```
I0613 17:07:29.500952       1 wrap.go:47] GET /api/v1/namespaces/kube-system/configmaps?fieldSelector=metadata.name%3Dfluentd-gcp-config-old-v1.2.5&resourceVersion=24085803&timeout=9m43s&timeoutSeconds=583&watch=true: (100.314899ms) 200 [kubelet/v1.16.0 (linux/amd64) kubernetes/4f29960 34.73.136.89:50938]
I0613 17:07:29.500955       1 wrap.go:47] GET /api/v1/namespaces/test-y8pkr4-2/secrets?fieldSelector=metadata.name%3Ddefault-token-t4m48&resourceVersion=6079445&timeout=8m22s&timeoutSeconds=502&watch=true: (100.416037ms) 200 [kubelet/v1.16.0 (linux/amd64) kubernetes/4f29960 34.73.136.89:50938]
I0613 17:07:29.500981       1 wrap.go:47] GET /api/v1/namespaces/test-y8pkr4-14/secrets?fieldSelector=metadata.name%3Ddefault-token-2xv9b&resourceVersion=6079445&timeout=9m39s&timeoutSeconds=579&watch=true: (142.428756ms) 200 [kubelet/v1.16.0 (linux/amd64) kubernetes/4f29960 34.73.136.89:50938]
I0613 17:07:29.501070       1 wrap.go:47] GET /api/v1/namespaces/test-y8pkr4-30/secrets?fieldSelector=metadata.name%3Ddefault-token-2jjwr&resourceVersion=6079445&timeout=5m41s&timeoutSeconds=341&watch=true: (100.963816ms) 200 [kubelet/v1.16.0 (linux/amd64) kubernetes/4f29960 34.73.136.89:50938]
I0613 17:07:29.501075       1 wrap.go:47] GET /api/v1/services?resourceVersion=6451001&timeoutSeconds=413&watch=true: (144.671946ms) 200 [kubelet/v1.16.0 (linux/amd64) kubernetes/4f29960 34.73.136.89:50938]
I0613 17:07:29.502456       1 wrap.go:47] GET /api/v1/namespaces/kube-system/secrets?fieldSelector=metadata.name%3Dmetadata-proxy-token-bsrjk&resourceVersion=6079445&timeout=7m46s&timeoutSeconds=466&watch=true: (102.372225ms) 200 [kubelet/v1.16.0 (linux/amd64) kubernetes/4f29960 34.73.136.89:50938]
```
From apiserver perspective the watches are being closed after ~0.1s. There is no error there.

**Comment by krzysied** [comment]
@alejandrox1 It looks like there is another regression in gce 5k, that needs to be investigated. Could you add milestone, please?

/cc @wojtek-t

**Comment by alejandrox1** [comment]
From the looks of it this looks like it is something to fix during v1.16, but please correct me if i'm wrong on this @krzysied 
/milestone v1.16

**Comment by mm4tt** [comment]
I'm not sure if we can move this to 1.16. 
The last run of the **release blocking** gce-scale-performance failed due to this:
![uF5GbcTyVVR](https://user-images.githubusercontent.com/2604887/59620862-c5612f00-90e2-11e9-83c1-bc081040a526.png)



@krzysied, have we tried running the test with reverted COS version upgrade? Last time we talked about this you had a hunch that it may be related.

**Comment by krzysied** [comment]
I'm also not sure about this 1.16. On the other hand is hard to make the decision when previous runs were affected by another regression...

Currently I'm running test to see if the problem is with COS. I'll post update tomorrow morning.

**Comment by alejandrox1** [comment]
@mm4tt @krzysied let me understand the scope of this better.

During code thaw (currently), these are the release blocking jobs [1.15-blocking](https://testgrid.k8s.io/sig-release-1.15-blocking).
During the rest of any release cycle, these are the release blocking jobs [master-blocking](https://testgrid.k8s.io/sig-release-master-blocking).
The ci-kubernetes-e2e-gce-scale-performance (the job listed in this issue) job is in the [master-informing](https://testgrid.k8s.io/sig-release-master-informing) dashboard which is not necessarily release blocking. 
First question should be, do you all consider this job should be release blocking?

Going back to this job, the aast run of ci-kubernetes-e2e-gce-scale-performance failed. For this run i see this new commits in k/k https://github.com/kubernetes/kubernetes/compare/4f29960cb...d8695d06b , all which I think came after code freeze was lifted. Do you think/know if the reason for this new failure is due to these new commits or is it something that was added during the 1.15 cycle or previous to it?

**Comment by mm4tt** [comment]
To recap the whole picture here:

1. On May 27th, in https://github.com/kubernetes/test-infra/pull/12773 we stopped overriding the cos version in our scalability test (we had done do it for some historical reasons that I only have a vague idea about).  Since then the scalability tests started using the oss default cos version, bumped up to cos-beta-73-11647-64-0 in https://github.com/kubernetes/kubernetes/pull/75149

2. Since May 28th, we've started seeing the symptoms described in this bug. Quoting @krzysied,

> There are additional services, pods, nodes calls made by kubelets. Test not always fails due to this, however those call have impact on other calls latency.

You can see it on the perf-dash graph from https://github.com/kubernetes/kubernetes/issues/79096#issuecomment-502606605
To shed some light on what happens, these requests are caused by kubelets that are randomly loosing connection to the api server. 
It doesn't make the test to fail constantly, but it makes the test flaky. That's one of the reasons we didn't catch the issue earlier. 

3. We didn't have any runs of gce-5k-perf in the first week of June due to prow issues - https://github.com/kubernetes/test-infra/issues/11594

4. Week ago, we've finally fixed the prow issues and found that we have another regression caused by the klog upgrade - https://github.com/kubernetes/kubernetes/issues/78734

5. After the culprit PR has been reverted, we've run into the regression described in this issue. The test failed which allowed us to discover it, but we've realize that symptoms are present since May 28th.

Our current theory is that the regression is caused by the cos version upgrade that happened on May 27th. If that happens to be true, then we have all the reasons to believe that the regression is also in 1.15, as it uses the newest cos version that we test in our master scalability tests.

@alejandrox1, that's why we believe that this regression may be release-blocking even though the gce-scale-regression doesn't run on 1.15 branch and technically it's not 1.15 blocking.

**Comment by claurence** [comment]
Thanks @mm4tt for the added details.

Given this we are delaying the 1.15 release to tomorrow to get a better idea of these failures and if we do need to hold the release for them.  Please let us know as early tomorrow as possible what the impact to release is (i.e. if we need to delay further than a day or if we can ship with these issues and fix them in a followup patch)

**Comment by mborsz** [comment]
https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/1140530001343418374 has failed as well with:

```
[measurement call APIResponsiveness - APIResponsiveness error: top latency metric: there should be no high-latency requests, but: [got: {Resource:replicationcontrollers Subresource: Verb:LIST Scope:cluster Latency:{Perc50:1.264ms Perc90:12.885261s Perc99:12.885261s} Count:4}; expected perc99 <= 10s]]
```

kube-apiserver.log for that request has this entry:
```
I0617 15:53:49.162675       1 trace.go:81] Trace[1942970888]: "List /api/v1/replicationcontrollers" (started: 2019-06-17 15:53:36.277436381 +0000 UTC m=+26921.715786951) (total time: 12.885206763s):
Trace[1942970888]: [470.026254ms] [469.986487ms] Listing from storage done
Trace[1942970888]: [12.885204014s] [12.41517776s] Writing http response done (16400 items)
I0617 15:53:49.162820       1 wrap.go:47] GET /api/v1/replicationcontrollers?labelSelector=group+%3D+load: (12.885696864s) 200 [clusterloader/v0.0.0 (linux/amd64) kubernetes/$Format 35.188.43.151:51052]
```

So we have spent most of the time in 'Writing http response' which can mean one of: lack of network or cpu in kube-apiserver or client, but based on our grafana dashboards for that period it looks like we have sufficient amount of both.

Another possibility is that serialization was slow that time (it's also happening inside of 'Writing http response' step), which can be caused by e.g. other big lists.

**Comment by mborsz** [comment]
I can't find any other LISTs around that time (there are some O(10s) earlier and after, but I don't think they can affect this given that they are finishing in O(80 ms)).

**Comment by mborsz** [comment]
clusterloader2 doesn't seem to be very busy that time:
```
W0617 15:53:48.590] I0617 15:53:48.590003   11833 wait_for_pods.go:139] WaitForControlledPodsRunning: namespace(test-41m4dw-44), labelSelector(name=small-rc-106): Pods: 2 out of 2 created, 2 running, 0 pending scheduled, 0 not scheduled, 0 inactive, 0 terminating, 0 unknown, 0 runn
ingButNotReady 
W0617 15:53:49.150] I0617 15:53:49.150168   11833 wait_for_pods.go:139] WaitForControlledPodsRunning: namespace(test-41m4dw-16), labelSelector(name=big-rc-0): Pods: 271 out of 271 created, 271 running, 0 pending scheduled, 0 not scheduled, 0 inactive, 0 terminating, 0 unknown, 0 runningButNotReady 
W0617 15:53:49.720] I0617 15:53:49.720355   11833 wait_for_pods.go:139] WaitForControlledPodsRunning: namespace(test-41m4dw-26), labelSelector(name=small-rc-241): Pods: 3 out of 3 created, 3 running, 0 pending scheduled, 0 not scheduled, 0 inactive, 0 terminating, 0 unknown, 0 runningButNotReady 
W0617 15:53:49.855] I0617 15:53:49.854764   11833 wait_for_pods.go:139] WaitForControlledPodsRunning: namespace(test-41m4dw-16), labelSelector(name=small-rc-265): Pods: 2 out of 2 created, 2 running, 0 pending scheduled, 0 not scheduled, 0 inactive, 0 terminating, 0 unknown, 0 runningButNotReady 
W0617 15:53:51.116] I0617 15:53:51.116221   11833 wait_for_pods.go:139] WaitForControlledPodsRunning: namespace(test-41m4dw-38), labelSelector(name=big-rc-0): Pods: 224 out of 224 created, 224 running, 0 pending scheduled, 0 not scheduled, 0 inactive, 0 terminating, 0 unknown, 0 runningButNotReady 
W0617 15:53:55.449] I0617 15:53:55.449648   11833 wait_for_controlled_pods.go:281] WaitForControlledPodsRunning: running 16400, deleted 0, timeout: 0, unknown: 0
W0617 15:53:55.453] I0617 15:53:55.450334   11833 wait_for_controlled_pods.go:295] WaitForControlledPodsRunning: 16400/16400 ReplicationControllers are running with all pods
W0617 15:53:55.454] I0617 15:53:55.450390   11833 simple_test_executor.go:126] Step "Scaling RCs" ended
W0617 15:54:00.643] I0617 15:54:00.643389   11833 wait_for_pods.go:139] WaitForControlledPodsRunning: namespace(test-41m4dw-7), labelSelector(name=small-rc-270): Pods: 0 out of 0 created, 0 running, 0 pending scheduled, 0 not scheduled, 0 inactive, 2 terminating, 0 unknown, 0 runningButNotReady 
```

**Comment by krzysied** [comment]
I've run tests with reverted COS. I haven't spotted any difference. There is still a problem with additional service lists.

**Comment by mborsz** [comment]
Re failing density test:

The increased latency is caused by a huge number of pods LISTS generated by node-controller (taint manager?).

On the initization, taint manager iterates all nodes and for each triggers /api/v1/pods?labelSelector=nodeName=X query which generates very big load on the cluster. This increases all the latencies.

In the 

In the 1140530001343418374 run, the latency graph looks like:
![xvdUn6X8SVj](https://user-images.githubusercontent.com/2559168/59679173-6f69b580-91ce-11e9-8079-cf1521787e61.png)

This perfectly correlates with queries generated by node-controller:
![5WQcR7hQUWr](https://user-images.githubusercontent.com/2559168/59681883-43056780-91d5-11e9-9cfa-5107a72cb1bd.png)

This happens also in green runs (e.g. the one from May 30th: 1134007042525106179), but usually finishes faster:
![j1doXjOmcuP](https://user-images.githubusercontent.com/2559168/59682002-8d86e400-91d5-11e9-8c52-d2898c7c4372.png)
(10:19 -- 10:34 is approx. 15 minutes, while the 1140530001343418374 run needed ~40 minutes to finish this.

This clearly can affect request latency in the initial phase of density test.

This is likely caused by the initial length of workqueue in taint manager:
1134007042525106179:
![LBx6naUzC90](https://user-images.githubusercontent.com/2559168/59682188-f8d0b600-91d5-11e9-9f0a-502a4b26d2f2.png)
1140530001343418374:
![wn7OPcxYJkL](https://user-images.githubusercontent.com/2559168/59682212-05eda500-91d6-11e9-899f-f114350407d7.png)

The workqueue in "May 30th" starts from 2.5K while in 1140530001343418374 it's more like 4K.

I was thinking that maybe in "May 30th" taint manager managed to do more work before prometheus started monitoring the workqueue depth, but it's not true:

Time difference between master creation and prometheus stack running  in 1140530001343418374 is:
```
W0617 08:24:05.144] gce-scale-cluster-master  us-east1-b  n1-standard-64               10.40.0.2    34.74.34.235  RUNNING
W0617 08:34:15.391] I0617 08:34:15.390169   11833 prometheus.go:118] Prometheus stack set up successfully
```

In May 30th:
```
W0530 08:09:14.233] gce-scale-cluster-master  us-east1-b  n1-standard-64               10.40.0.2    34.74.174.63  RUNNING
W0530 08:19:42.753] I0530 08:19:42.751878   12200 prometheus.go:118] Prometheus stack set up successfully
```

**Comment by mborsz** [comment]
For me, it looks like we have three different issues right now:
1. Density test -- node controller overloading master
2. replicaitoncontrolller LIST latency in load
3. Increased number of LIST services in load test.

While 2. can be a flake, 1. and 3. both seem to be valid issues.

**Comment by claurence** [comment]
@wojtek-t @shyamjvs - have you all gotten enough information to determine if this should block the 1.15 release?

**Comment by mborsz** [comment]
Just to dump our (my and @krzysied) understanding of the problem '3.'.

For some unknown yet reason, kubelets are failing to renew lease:

e.g. in 1140530001343418374, ```104.196.17.9``` node:
```
E0617 18:53:55.074203    1736 controller.go:149] failed to update node lease, error: Put https://34.74.34.235/apis/coordination.k8s.io/v1beta1/namespaces/kube-node-lease/leases/gce-scale-cluster-minion-group-2c696?timeout=10s: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
E0617 18:53:59.125451    1736 controller.go:149] failed to update node lease, error: Operation cannot be fulfilled on leases.coordination.k8s.io "gce-scale-cluster-minion-group-2-c696": the object has been modified; please apply your changes to the latest version and try again
E0617 18:53:59.189452    1736 controller.go:149] failed to update node lease, error: Operation cannot be fulfilled on leases.coordination.k8s.io "gce-scale-cluster-minion-group-2-c696": the object has been modified; please apply your changes to the latest version and try again
W0617 18:53:59.189997    1736 reflector.go:306] object-"kube-system"/"fluentd-gcp-config-old-v1.2.5": watch of *v1.ConfigMap ended with: very short watch: object-"kube-system"/"fluentd-gcp-config-old-v1.2.5": Unexpected watch close - watch lasted less than a second and no items received
W0617 18:53:59.191432    1736 reflector.go:306] k8s.io/kubernetes/pkg/kubelet/kubelet.go:444: watch of *v1.Service ended with: very short watch: k8s.io/kubernetes/pkg/kubelet/kubelet.go:444: Unexpected watch close - watch lasted less than a second and no items received
W0617 18:53:59.191654    1736 reflector.go:306] k8s.io/client-go/informers/factory.go:133: watch of *v1beta1.RuntimeClass ended with: very short watch: k8s.io/client-go/informers/factory.go:133: Unexpected watch close - watch lasted less than a second and no items received
W0617 18:53:59.191850    1736 reflector.go:306] k8s.io/client-go/informers/factory.go:133: watch of *v1beta1.CSIDriver ended with: very short watch: k8s.io/client-go/informers/factory.go:133: Unexpected watch close - watch lasted less than a second and no items received
W0617 18:53:59.192033    1736 reflector.go:306] object-"kube-system"/"metadata-proxy-token-s8jk7": watch of *v1.Secret ended with: very short watch: object-"kube-system"/"metadata-proxy-token-s8jk7": Unexpected watch close - watch lasted less than a second and no items received
W0617 18:53:59.193370    1736 reflector.go:306] k8s.io/kubernetes/pkg/kubelet/config/apiserver.go:47: watch of *v1.Pod ended with: very short watch: k8s.io/kubernetes/pkg/kubelet/config/apiserver.go:47: Unexpected watch close - watch lasted less than a second and no items received
W0617 18:53:59.193586    1736 reflector.go:306] object-"test-41m4dw-27"/"default-token-d89fh": watch of *v1.Secret ended with: very short watch: object-"test-41m4dw-27"/"default-token-d89fh": Unexpected watch close - watch lasted less than a second and no items received
W0617 18:53:59.193781    1736 reflector.go:306] k8s.io/kubernetes/pkg/kubelet/kubelet.go:453: watch of *v1.Node ended with: very short watch: k8s.io/kubernetes/pkg/kubelet/kubelet.go:453: Unexpected watch close - watch lasted less than a second and no items received
W0617 18:53:59.193959    1736 reflector.go:306] object-"kube-system"/"fluentd-gcp-token-z8mrs": watch of *v1.Secret ended with: very short watch: object-"kube-system"/"fluentd-gcp-token-z8mrs": Unexpected watch close - watch lasted less than a second and no items received
W0617 18:53:59.195935    1736 reflector.go:306] object-"test-41m4dw-47"/"default-token-qnrxf": watch of *v1.Secret ended with: very short watch: object-"test-41m4dw-47"/"default-token-qnrxf": Unexpected watch close - watch lasted less than a second and no items received
W0617 18:53:59.196222    1736 reflector.go:306] object-"test-41m4dw-44"/"default-token-dlcc9": watch of *v1.Secret ended with: very short watch: object-"test-41m4dw-44"/"default-token-dlcc9": Unexpected watch close - watch lasted less than a second and no items received
W0617 18:53:59.196761    1736 reflector.go:306] k8s.io/kubernetes/pkg/kubelet/kubeletconfig/controller.go:227: watch of *v1.Node ended with: very short watch: k8s.io/kubernetes/pkg/kubelet/kubeletconfig/controller.go:227: Unexpected watch close - watch lasted less than a second and no items received
E0617 18:53:59.206168    1736 controller.go:149] failed to update node lease, error: Operation cannot be fulfilled on leases.coordination.k8s.io "gce-scale-cluster-minion-group-2-c696": the object has been modified; please apply your changes to the latest version and try again
E0617 18:53:59.214819    1736 controller.go:149] failed to update node lease, error: Operation cannot be fulfilled on leases.coordination.k8s.io "gce-scale-cluster-minion-group-2-c696": the object has been modified; please apply your changes to the latest version and try again
```

What is interesting is that although kubelet claims it hasn't received response to PUT:

```
E0617 18:53:55.074203    1736 controller.go:149] failed to update node lease, error: Put https://34.74.34.235/apis/coordination.k8s.io/v1beta1/namespaces/kube-node-lease/leases/gce-scale-cluster-minion-group-2-c696?timeout=10s: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
```
kube-apiserver.log says it finished in 7ms:
```
I0617 18:53:45.081222       1 wrap.go:47] PUT /apis/coordination.k8s.io/v1beta1/namespaces/kube-node-lease/leases/gce-scale-cluster-minion-group-2-c696?timeout=10s: (7.183655ms) 200 [kubelet/v1.16.0 (linux/amd64) kubernetes/8f4cc7d 104.196.17.9:60132]
```

Then kubelet retries to PUT update which fails as the previous put has succeeded and after few retries it calls onRepeatedHeartbeatFailure(https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/nodelease/controller.go#L151) which is set to 'closeAllConn' function (https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L574).

Closing connection interrupts all watches which then creates LIST service (and other lists as well).

So this is "unable to renew lease" => "connection closed" reasoning and not otherwise as we thought initially.

The question why we are not able to update lease still holds, I'm trying to rerun the test with GODEBUG=http2client=2 in kubelet to understand at what level the response is dropped.

Another thing is that it looks like attempt to retry 409 error is quite unlikely to be successful and will close all connections. This is a bug that should be fixed at some point.

**Comment by claurence** [comment]
@mborsz would anyone who is looking into this issue be able to join our 1.15 release call in an hour (8am PT) to give an update?

**Comment by mm4tt** [comment]
Thanks @mborsz and @krzysied, great analysis!

If we can exclude the COS version upgrade, then there is a big chance that the regression is only in the master branch and not in the 1.15 branch. 
We'll need one more day though to verify this assumption. 

We'll be running the tests shortly on 1.15, they take ~12h so we'll have the results by tomorrow's morning.

**Comment by mborsz** [comment]
The 1.15 test has finished with the following result:
1) in load test latency of replicationcontroller LIST is 15s, so it's out of SLO.
2) there is a single service LIST caused by failure of physical machine running one of nodes, so this is not this issue (it hasn't changed the test result).
3) density test has passed

For 1) I think we can assume that it's flake (we have only 4 probes of replicationcontroller LISTs in load test and we started measuring this metric recently).
For 2) Given that in each run of 5000 node test since 2019-05-28 (when the issue started) we saw [at least ~20 service LISTs caused by this issue](http://perf-dash.k8s.io/#/?jobname=gce-5000Nodes&metriccategoryname=APIServer&metricname=LoadRequestCountByClient&client=kubelet%2Fv1.16.0%20(linux%2Famd64)%20kubernetes&code=200&component=apiserver&group=&resource=services&scope=cluster&subresource=&verb=LIST&version=v1) and here we have 0, it looks like 1.15 branch is not affected by this issue.

Based on that I think it's OK to unblock 1.15 release.

We will still work on fixing this issue.

**Comment by mm4tt** [comment]
Great job, @mborsz and @krzysied!

Thank you for all your hard work that led to unblocking the 1.15 release!

**Comment by krzysied** [comment]
When additional calls started to happen, there was also huge cpu usage drop in the main components:
 
[apiserver cpu usage](http://perf-dash.k8s.io/#/?jobname=gce-5000Nodes&metriccategoryname=E2E&metricname=LoadResources&PodName=kube-apiserver-gce-scale-cluster-master%2Fkube-apiserver&Resource=CPU)

[scheduler cpu usage](http://perf-dash.k8s.io/#/?jobname=gce-5000Nodes&metriccategoryname=E2E&metricname=LoadResources&PodName=kube-scheduler-gce-scale-cluster-master%2Fkube-scheduler&Resource=CPU)

[etcd cpu usage](http://perf-dash.k8s.io/#/?jobname=gce-5000Nodes&metriccategoryname=E2E&metricname=LoadResources&PodName=etcd-server-gce-scale-cluster-master%2Fetcd-container&Resource=CPU)

**Comment by krzysied** [comment]
I've added separate issue for increased number of RC/status update calls - #79277. This regression also started on 2019-05-28 so it can be somehow connected.

**Comment by krzysied** [comment]
Regarding density test and difference between runs 1134007042525106179 and 1140530001343418374...

Run 1134007042525106179: 5004 pods listing calls where made during initialization. 5001 of them were unique.
Run 1140530001343418374 5506 pods listing calls where made during initialization, 5001 of them were unique.

Apiserver handles ~50 pod list per minute, so these 500 additional call convert to 10 additional minutes.

**Comment by krzysied** [comment]
Apisever logs
```
I0617 08:26:45.542296       1 wrap.go:47] GET /api/v1/pods?fieldSelector=spec.nodeName%3Dgce-scale-cluster-minion-group-1-019l: (3.034157ms) 200 [kube-controller-manager/v1.16.0 (linux/amd64) kubernetes/8f4cc7d/system:serviceaccount:kube-system:node-controller [::1]:49332]
I0617 08:30:41.254234       1 wrap.go:47] GET /api/v1/pods?fieldSelector=spec.nodeName%3Dgce-scale-cluster-minion-group-1-019l: (1.601604214s) 200 [kube-controller-manager/v1.16.0 (linux/amd64) kubernetes/8f4cc7d/system:serviceaccount:kube-system:node-controller [::1]:50990]
```

[Node logs](https://gcsweb.k8s.io/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/1140530001343418374/artifacts/gce-scale-cluster-minion-group-1-019l/)
[kube-controller-manager logs](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/1140530001343418374/artifacts/gce-scale-cluster-master/kube-controller-manager.log)

I haven't found anything suspicious in kubelet logs.

Kube-controller-manager:

```
I0617 08:26:47.254424       1 node_lifecycle_controller.go:644] Controller observed a new Node: "gce-scale-cluster-minion-group-1-019l"
I0617 08:26:47.254449       1 controller_utils.go:168] Recording Registered Node gce-scale-cluster-minion-group-1-019l in Controller event message for node gce-scale-cluster-minion-group-1-019l
I0617 08:26:47.254532       1 event.go:258] Event(v1.ObjectReference{Kind:"Node", Namespace:"", Name:"gce-scale-cluster-minion-group-1-019l", UID:"bd20bb42-c248-4844-961b-af87904f93e1", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'RegisteredNode' Node 
gce-scale-cluster-minion-group-1-019l event: Registered Node gce-scale-cluster-minion-group-1-019l in Controller
W0617 08:26:47.254703       1 node_lifecycle_controller.go:833] Missing timestamp for Node gce-scale-cluster-minion-group-1-019l. Assuming now as a timestamp.
I0617 08:26:57.257561       1 service_controller.go:643] Detected change in list of current cluster nodes. New node set: map[gce-scale-cluster-minion-group-1-019l:{} gce-scale-cluster-minion-group-1-0cff:{} gce-scale-cluster-minion-group-1-0m4k:{} gce-scale-cluster-minion-group-1-2
15d:{} gce-scale-cluster-minion-group-1-3cvn:{} gce-scale-cluster-minion-group-1-4fc8:{} gce-scale-cluster-minion-group-1-7kl2:{} gce-scale-cluster-minion-group-1-8977:{} gce-scale-cluster-minion-group-1-8l7g:{} gce-scale-cluster-minion-group-1-b50g:{} gce-scale-cluster-minion-grou
p-1-d893:{} gce-scale-cluster-minion-group-1-f29l:{} gce-scale-cluster-minion-group-1-hjhs:{} gce-scale-cluster-minion-group-1-j42x:{} gce-scale-cluster-minion-group-1-jbx6:{} gce-scale-cluster-minion-group-1-jvwx:{} gce-scale-cluster-minion-group-1-k6k6:{} gce-scale-cluster-minion
-group-1-kmcj:{} gce-scale-cluster-minion-group-1-kxvr:{} gce-scale-cluster-minion-group-1-l3lf:{} gce-scale-cluster-minion-group-1-lcxx:{} gce-scale-cluster-minion-group-1-lq5h:{} gce-scale-cluster-minion-group-1-lvkq:{} gce-scale-cluster-minion-group-1-mcs5:{} gce-scale-cluster-m
inion-group-1-mj43:{} gce-scale-cluster-minion-group-1-nd05:{} gce-scale-cluster-minion-group-1-nm23:{} gce-scale-cluster-minion-group-1-nmr4:{} gce-scale-cluster-minion-group-1-qwh0:{} gce-scale-cluster-minion-group-1-srn9:{} gce-scale-cluster-minion-group-1-vq56:{} gce-scale-clus
ter-minion-group-1-w4xp:{} gce-scale-cluster-minion-group-1-w87n:{} gce-scale-cluster-minion-group-1-w95m:{} gce-scale-cluster-minion-group-1-xlwt:{} gce-scale-cluster-minion-group-1-z7ps:{} gce-scale-cluster-minion-group-1-zp8m:{} gce-scale-cluster-minion-heapster:{}]
I0617 08:27:17.227150       1 ttl_controller.go:271] Changed ttl annotation for node gce-scale-cluster-minion-group-1-019l to 60 seconds
W0617 08:28:42.990200       1 actual_state_of_world.go:506] Failed to update statusUpdateNeeded field in actual state of world: Failed to set statusUpdateNeeded to needed true, because nodeName="gce-scale-cluster-minion-group-1-019l" does not exist

```

```
I0617 08:28:43.907626       1 node_lifecycle_controller.go:644] Controller observed a new Node: "gce-scale-cluster-minion-group-1-019l"
I0617 08:28:43.907632       1 controller_utils.go:168] Recording Registered Node gce-scale-cluster-minion-group-1-019l in Controller event message for node gce-scale-cluster-minion-group-1-019l
I0617 08:28:43.908064       1 event.go:258] Event(v1.ObjectReference{Kind:"Node", Namespace:"", Name:"gce-scale-cluster-minion-group-1-019l", UID:"bd20bb42-c248-4844-961b-af87904f93e1", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'RegisteredNode' Node gce-scale-cluster-minion-group-1-019l event: Registered Node gce-scale-cluster-minion-group-1-019l in Controller
W0617 08:28:43.972474       1 node_lifecycle_controller.go:833] Missing timestamp for Node gce-scale-cluster-minion-group-1-019l. Assuming now as a timestamp.
I0617 08:29:30.702166       1 ttl_controller.go:271] Changed ttl annotation for node gce-scale-cluster-minion-group-1-019l to 300 seconds
```

**Comment by krzysied** [comment]
Line ```W0617 08:28:42.990200       1 actual_state_of_world.go:506] Failed to update statusUpdateNeeded field in actual state of world: Failed to set statusUpdateNeeded to needed true, because nodeName="gce-scale-cluster-minion-group-1-019l" does not exist``` seems like a issue here, however I'm not sure what caused it...

**Comment by krzysied** [comment]
I looked into controller code. There is no way to re-register node in node_lifecycle_controller without deleting node first or restarting the controller-manager.

Then I've found in the logs:
```
F0617 08:27:32.551398       1 controllermanager.go:281] leaderelection lost
I0617 08:28:14.546523       1 feature_gate.go:216] feature gates: &{map[ExperimentalCriticalPodAnnotation:true]}
I0617 08:28:14.546679       1 flags.go:33] FLAG: --address="0.0.0.0"
I0617 08:28:14.546694       1 flags.go:33] FLAG: --allocate-node-cidrs="true"
I0617 08:28:14.546704       1 flags.go:33] FLAG: --allow-untagged-cloud="false"
I0617 08:28:14.546712       1 flags.go:33] FLAG: --alsologtostderr="false"
```

which explains the additional node adds.

**Comment by krzysied** [comment]
Investigating load test latency of replicationcontroller LIST will be handled in separate issue - #79596

**Comment by wojtek-t** [comment]
Apart from #79596 everything else from this issue has been fixed. Closing this one.
